glean-core/src/lib.rs

```diff
@@ -128,6 +170,76 @@ impl Glean {
         self.upload_enabled
     }
 
+    /// Handles the changing of state when upload_enabled changes.
+    ///
+    /// Should only be called when the state actually changes.
+    /// When disabling, all pending metrics, events and queued pings are cleared.
+    ///
+    /// When enabling, the core Glean metrics are recreated.
+    ///
+    /// # Arguments
+    ///
+    /// * `flag` - When true, enable metric collection.
+    fn on_change_upload_enabled(&mut self, flag: bool) {
+        if flag {
+            self.initialize_core_metrics();
+        } else {
+            self.clear_metrics();
+        }
+    }
+
+    /// Clear any pending metrics when telemetry is disabled.
+    fn clear_metrics(&mut self) {
+        // There is only one metric that we want to survive after clearing all
+        // metrics: first_run_date. Here, we store its value so we can restore
+        // it after clearing the metrics.
+        let existing_first_run_date = self
+            .core_metrics
+            .first_run_date
+            .get_value(self, "glean_client_info");
+
+        // Clear any pending pings.
+        let ping_maker = PingMaker::new();
+        if let Err(err) = ping_maker.clear_pending_pings(self.get_data_path()) {
+            log::error!("Error clearing pending pings: {}", err);
+        }
+
+        // Delete all stored metrics.
+        // Note that this also includes the ping sequence numbers, so it has
+        // the effect of resetting those to their initial values.
+        self.data_store.clear_all();
+        // TODO: 1552872: Clear the event store, once the event metric lands.
```

Comment:
Since you're working on both the PRs, let's make sure this either happens here or in #161 (so that it doesn't get lost). You could file a follow-up bug, if you think is appropriate.

Reply:
I plan to merge #161 first, then this can be added here.